File: android/src/main/java/com/leokinder2k/koratuningcompanion/notation/engine/StringId.kt
package com.leokinder2k.koratuningcompanion.notation.engine

// Port of stringId.js

data class ParsedStringId(
    val side: Char,                    // 'L' or 'R'
    val physicalIndexFromGourd: Int,   // 1-based
)

fun parseStringId(stringId: String): ParsedStringId {
    require(stringId.length >= 2) { "Invalid stringId: $stringId" }
    val side = stringId[0]
    require(side == 'L' || side == 'R') { "Invalid stringId side: $stringId" }
    val num = stringId.substring(1).toIntOrNull()
        ?: error("Invalid stringId index: $stringId")
    require(if (side == 'R') num >= 0 else num >= 1) { "Invalid stringId index: $stringId" }
    return ParsedStringId(side, num)
}

fun allStringIds(instrumentType: KoraInstrumentType): List<String> {
    val ids = mutableListOf<String>()
    for (i in 1..instrumentType.leftCount) ids.add("L$i")
    val firstRight = if (instrumentType == KoraInstrumentType.KORA_22_CHROMATIC) 0 else 1
    for (i in firstRight until firstRight + instrumentType.rightCount) ids.add("R$i")
    return ids
}

/**
 * Thumb numbering starts at the bass end: the lowest string in the thumb zone is 1.
 * Finger numbering starts at the treble end: the highest string in the finger zone is 1.
 */
fun renderedNumber(instrumentType: KoraInstrumentType, stringId: String, digitLine: String): Int {
    val (side, displayIndex) = parseStringId(stringId)
    val n = if (side == 'L') instrumentType.leftCount else instrumentType.rightCount
    val physicalIndex = if (side == 'R' && instrumentType == KoraInstrumentType.KORA_22_CHROMATIC) {
        displayIndex + 1
    } else {
        displayIndex
    }
    require(physicalIndex in 1..n) { "Invalid stringId index for instrument: $stringId" }
    val usesFingerNumbering = digitLine == "LF" || digitLine == "RF"
    val usesThumbNumbering = digitLine == "LT" || digitLine == "RT"
    require(usesFingerNumbering || usesThumbNumbering) { "Unknown digitLine: $digitLine" }
    return if (usesThumbNumbering) {
        require(physicalIndex <= MAX_TAB_DIGIT) {
            "$stringId is outside the six-string thumb zone for $digitLine"
        }
        physicalIndex
    } else {
        val firstFingerIndex = maxOf(1, n - MAX_TAB_DIGIT + 1)
        require(physicalIndex >= firstFingerIndex) {
            "$stringId is outside the six-string finger zone for $digitLine"
        }
        n - physicalIndex + 1
    }
}

private const val MAX_TAB_DIGIT = 6

File: apple/kora_engine/stringId.js
import { instrumentModel } from "./instrument.js";

export const Side = Object.freeze({
  L: "L",
  R: "R",
});

/**
 * Canonical IDs:
 * - Left:  L01..L11
 * - Right: R01..R10 or R01..R11
 */
export function parseStringId(stringId) {
  if (typeof stringId !== "string" || stringId.length < 3) {
    throw new Error(`Invalid stringId: ${stringId}`);
  }
  const side = stringId[0];
  if (side !== "L" && side !== "R") throw new Error(`Invalid stringId side: ${stringId}`);
  const num = Number.parseInt(stringId.slice(1), 10);
  if (!Number.isInteger(num) || num < 1) throw new Error(`Invalid stringId index: ${stringId}`);
  return { side, physicalIndexFromGourd: num };
}

export function allStringIds(instrumentType) {
  const { leftCount, rightCount } = instrumentModel(instrumentType);
  const ids = [];
  for (let i = 1; i <= leftCount; i++) ids.push(`L${String(i).padStart(2, "0")}`);
  for (let i = 1; i <= rightCount; i++) ids.push(`R${String(i).padStart(2, "0")}`);
  return ids;
}

/**
 * Finger numbering is physical index. Thumb numbering is reversed (N+1-index).
 */
export function renderedNumber({ instrumentType, stringId, digitLine }) {
  const { side, physicalIndexFromGourd } = parseStringId(stringId);
  const { leftCount, rightCount } = instrumentModel(instrumentType);
  const N = side === "L" ? leftCount : rightCount;

  const usesFingerNumbering = digitLine === "LF" || digitLine === "RF";
  const usesThumbNumbering = digitLine === "LT" || digitLine === "RT";
  if (!usesFingerNumbering && !usesThumbNumbering) throw new Error(`Unknown digitLine: ${digitLine}`);

  if (usesFingerNumbering) return physicalIndexFromGourd;
  return N + 1 - physicalIndexFromGourd;
}

